File: BlackFireIM/Example/Home/View/BFSparkCardCell.m
//
//  BFSparkCardCell.m
//  BlackFireIM
//
//  Created by benny wang on 2021/4/13.
//

#import "BFSparkCardCell.h"
#import "MSIMSDK-UIKit.h"
#import <MSIMSDK/MSIMSDK.h>
#import <SDWebImage.h>


@interface BFSparkCardCell()

@property(nonatomic,strong) CAGradientLayer *gradientLayer;

@property(nonatomic,strong) MSProfileInfo *user;

@end
@implementation BFSparkCardCell

- (instancetype)initWithReuseIdentifier:(NSString *)reuseIdentifier
{
    if (self = [super initWithReuseIdentifier:reuseIdentifier]) {
        
        self.backgroundColor = [UIColor whiteColor];
        self.imageView = [[UIImageView alloc]init];
        self.imageView.contentMode = UIViewContentModeScaleAspectFill;
        self.imageView.userInteractionEnabled = YES;
        [self addSubview:self.imageView];
        
        self.gradientLayer = [CAGradientLayer layer];
        self.gradientLayer.colors = @[(id)RGBA(0, 0, 0, 0.0).CGColor,(id)RGBA(0, 0, 0, 0.8).CGColor];
        self.gradientLayer.startPoint = CGPointMake(0, 0);
        self.gradientLayer.endPoint = CGPointMake(0, 1);
        [self.imageView.layer addSublayer:self.gradientLayer];
        
        self.title = [[UILabel alloc]init];
        self.title.textColor = [UIColor whiteColor];
        self.title.font = [UIFont boldSystemFontOfSize:18.0f];
        [self.imageView addSubview:self.title];
        
        self.dislike = [[UIImageView alloc]init];
        self.dislike.image = [UIImage imageNamed:@"finder_dislike_btn"];
        self.dislike.alpha = 0.0f;
        [self.imageView addSubview:self.dislike];
        
        self.like = [[UIImageView alloc]init];
        self.like.image = [UIImage imageNamed:@"finder_like_btn"];
        self.like.alpha = 0.0f;
        [self.imageView addSubview:self.like];

        self.winkBtn = [UIButton buttonWithType:UIButtonTypeCustom];
        [self.winkBtn setImage:[UIImage imageNamed:@"spark_wink"] forState:UIControlStateNormal];
        [self.winkBtn setImage:[UIImage imageNamed:@"spark_wink_sel"] forState:UIControlStateSelected];
        [self.winkBtn setImage:[UIImage imageNamed:@"spark_wink_sel"] forState:UIControlStateSelected | UIControlStateHighlighted];
        [self.winkBtn addTarget:self action:@selector(winkBtnDidClick) forControlEvents:UIControlEventTouchUpInside];
        [self.imageView addSubview:self.winkBtn];
        
        self.chatBtn = [UIButton buttonWithType:UIButtonTypeCustom];
        [self.chatBtn setImage:[UIImage imageNamed:@"spark_chat"] forState:UIControlStateNormal];
        [self.chatBtn addTarget:self action:@selector(chatBtnDidClick) forControlEvents:UIControlEventTouchUpInside];
        [self.imageView addSubview:self.chatBtn];
        
        self.layer.cornerRadius = 10;
        self.layer.borderColor = [UIColor lightGrayColor].CGColor;
        self.layer.borderWidth = 1;
        self.clipsToBounds = YES;
    }
    return self;
}

- (void)configItem:(MSProfileInfo *)info
{
    _user = info;
    [self.imageView sd_setImageWithURL:[NSURL URLWithString:info.pic] placeholderImage:nil options:SDWebImageRetryFailed | SDWebImageAllowInvalidSSLCertificates];
    self.title.text = info.nick_name;
    self.winkBtn.selected = NO;
}

- (void)winkBtnDidClick
{
    if ([self.delegate respondsToSelector:@selector(winkBtnDidClick:)]) {
        [self.delegate winkBtnDidClick:self];
    }
    [UIDevice impactFeedback];
}

- (void)chatBtnDidClick
{
    if ([self.delegate respondsToSelector:@selector(chatBtnDidClick:)]) {
        [self.delegate chatBtnDidClick:self];
    }
    [UIDevice impactFeedback];
}

- (void)layoutSubviews
{
    [super layoutSubviews];
    self.imageView.frame = self.bounds;
    self.gradientLayer.frame = CGRectMake(0, self.imageView.height-250, self.imageView.width, 250);
    self.title.frame = CGRectMake(20,self.height-120, 200, 30);
    self.like.frame = CGRectMake(16, 16, 75, 75);
    self.dislike.frame = CGRectMake(self.frame.size.width-21-75, 16, 75, 75);
    self.winkBtn.frame = CGRectMake(20, self.title.maxY+15, 50, 50);
    self.chatBtn.frame = CGRectMake(self.winkBtn.maxX+15, self.winkBtn.y, self.winkBtn.width, self.winkBtn.height);
}

@end

File: Swift/MSIM-Swift/Home/BFSparkCardCell.swift
//
//  BFSparkCardCell.swift
//  MSIM-Swift
//
//  Created by benny wang on 2021/8/2.
//

import UIKit
import MSIMSDK
import Kingfisher

protocol BFSparkCardCellDelegate: NSObjectProtocol {
    
    func winkBtnDidClick(cell: BFSparkCardCell)
    
    func chatBtnDidClick(cell: BFSparkCardCell)
}

class BFSparkCardCell: BFCardViewCell {

    var delegate: BFSparkCardCellDelegate?
    
    var titleL: UILabel!
    
    var imageView: UIImageView!
    
    var dislike: UIImageView!
    
    var like: UIImageView!
    
    var chatBtn: UIButton!
    
    var winkBtn: UIButton!
    
    var gradientLayer: CAGradientLayer!
    
    var profile: MSProfileInfo?
    
    override init(reuseIdentifier: String) {
        super.init(reuseIdentifier: reuseIdentifier)
        
        backgroundColor = .white
        imageView = UIImageView()
        imageView.contentMode = .scaleAspectFill
        imageView.isUserInteractionEnabled = true
        addSubview(imageView)
        
        gradientLayer = CAGradientLayer()
        gradientLayer.colors = [UIColor(r: 0, g: 0, b: 0, alpha: 0).cgColor,UIColor(r: 0, g: 0, b: 0, alpha: 0.8).cgColor]
        gradientLayer.startPoint = .zero
        gradientLayer.endPoint = CGPoint(x: 0, y: 1)
        imageView.layer.addSublayer(gradientLayer)
        
        titleL = UILabel()
        titleL.textColor = .white
        titleL.font = .boldSystemFont(ofSize: 18)
        imageView.addSubview(titleL)
        
        dislike = UIImageView()
        dislike.image = UIImage(named: "finder_dislike_btn")
        dislike.alpha = 0
        imageView.addSubview(dislike)
        
        like = UIImageView()
        like.image = UIImage(named: "finder_like_btn")
        like.alpha = 0
        imageView.addSubview(like)
        
        winkBtn = UIButton(type: .custom)
        winkBtn.setImage(UIImage(named: "spark_wink"), for: .normal)
        winkBtn.setImage(UIImage(named: "spark_wink_sel"), for: .selected)
        winkBtn.setImage(UIImage(named: "spark_wink_sel"), for: [.selected,.highlighted])
        winkBtn.addTarget(self, action: #selector(winkBtnDidClick), for: .touchUpInside)
        imageView.addSubview(winkBtn)
        
        chatBtn = UIButton(type: .custom)
        chatBtn.setImage(UIImage(named: "spark_chat"), for: .normal)
        chatBtn.addTarget(self, action: #selector(chatBtnDidClick), for: .touchUpInside)
        imageView.addSubview(chatBtn)
        
        layer.cornerRadius = 10
        layer.borderColor = UIColor.lightGray.cgColor
        layer.borderWidth = 1
        clipsToBounds = true
    }
    
    func configItem(item: MSProfileInfo) {
        
        self.profile = item
        imageView.kf.setImage(with: URL(string: item.pic))
        titleL.text = item.nick_name
        winkBtn.isSelected = false
    }
    
    @objc func winkBtnDidClick() {
        delegate?.winkBtnDidClick(cell: self)
        UIDevice.impactFeedback()
    }
    
    @objc func chatBtnDidClick() {
        delegate?.chatBtnDidClick(cell: self)
        UIDevice.impactFeedback()
    }
    
    override func layoutSubviews() {
        super.layoutSubviews()
        imageView.frame = bounds
        gradientLayer.frame = CGRect(x: 0, y: imageView.height - 250, width: imageView.width, height: 250)
        titleL.frame = CGRect(x: 20, y: height - 120, width: 200, height: 30)
        like.frame = CGRect(x: 16, y: 16, width: 75, height: 75)
        dislike.frame = CGRect(x: width - 21 - 75, y: 16, width: 75, height: 75)
        winkBtn.frame = CGRect(x: 20, y: titleL.bottom + 15, width: 50, height: 50)
        chatBtn.frame = CGRect(x: winkBtn.right + 15, y: winkBtn.top, width: winkBtn.width, height: winkBtn.height)
    }
    
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
}
